[SPARK-4286] Integrate external shuffle service to coarse grained Mesos mode

tnachen opened the pull request.

tnachen committed:
Launch External Shuffle Service with mesos

tnachen committed:
Support total and kill executors in coarse grained mesos mode.

tnachen committed:
Propogate the shuffle service setting.

tnachen committed:
Implement a new executor for coarse grained mesos mode.

tnachen committed:
Launch executor with shell and add traces

tnachen committed:
Fix destroying executor.

tnachen committed:
Only process status update if task is still tracked.

tnachen committed:
Generate new spark executor id on each launch and fix tests

SparkQA commented:
[Test build #24976 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/24976/consoleFull) for   PR 3861 at commit [`392486e`](https://github.com/apache/spark/commit/392486e0e6abb913bd1eb6a744825823e584c6fc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/24976/
Test FAILed.

tnachen commented:
I don't think the test failure is related, and the test passes locally too.

ash211 commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -998,7 +998,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   override def requestExecutors(numAdditionalExecutors: Int): Boolean = {
-    assert(master.contains("yarn") || dynamicAllocationTesting,
+    assert(master.contains("mesos") || master.contains("yarn") || dynamicAllocationTesting,
       "Requesting executors is currently only supported in YARN mode")
```
Change this message to be "... only supported in YARN or Mesos modes", and the message below

ash211 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+      frameworkInfo: FrameworkInfo,
+      slaveInfo: SlaveInfo) {
+    this.driver = driver
+    logInfo("Coarse Grain Mesos Executor '" + executorInfo.getExecutorId.getValue +
```
Grained

tnachen committed:
Address review feedback

tnachen commented:
@ash211 Thanks for the review, updated the PR.

SparkQA commented:
[Test build #25002 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25002/consoleFull) for   PR 3861 at commit [`a8d036c`](https://github.com/apache/spark/commit/a8d036cf6ec4b8b1fa621a4da955f3274517e41f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25002 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25002/consoleFull) for   PR 3861 at commit [`a8d036c`](https://github.com/apache/spark/commit/a8d036cf6ec4b8b1fa621a4da955f3274517e41f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25002/
Test PASSed.

tnachen commented:
@aarondav

tnachen committed:
Add documentation for new config added

SparkQA commented:
[Test build #25082 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25082/consoleFull) for   PR 3861 at commit [`ae1d91a`](https://github.com/apache/spark/commit/ae1d91a13f0e3bf91a6347dbfc0fa8ae7011674a).
 * This patch merges cleanly.

tnachen committed:
Fix documentation

SparkQA commented:
[Test build #25083 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25083/consoleFull) for   PR 3861 at commit [`c601c3f`](https://github.com/apache/spark/commit/c601c3f092e68c7553e1daa6a3648206a537e250).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25082 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25082/consoleFull) for   PR 3861 at commit [`ae1d91a`](https://github.com/apache/spark/commit/ae1d91a13f0e3bf91a6347dbfc0fa8ae7011674a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`
  * `    implicit class StringToAttributeConversionHelper(val sc: StringContext) `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25082/
Test PASSed.

SparkQA commented:
[Test build #25083 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25083/consoleFull) for   PR 3861 at commit [`c601c3f`](https://github.com/apache/spark/commit/c601c3f092e68c7553e1daa6a3648206a537e250).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25083/
Test PASSed.

SparkQA commented:
[Test build #25119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25119/consoleFull) for   PR 3861 at commit [`c954dc7`](https://github.com/apache/spark/commit/c954dc7d8a4bef3a76a43f14cfe1de6a0e3e440a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25119/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/StandaloneWorkerShuffleService.scala:
```diff
@@ -54,6 +53,8 @@ class StandaloneWorkerShuffleService(sparkConf: SparkConf, securityManager: Secu
       logInfo(s"Starting shuffle service on port $port with useSasl = $useSasl")
       server = transportContext.createServer(port)
+    } else {
+      logInfo(s"Skip launching shuffle service as it's not enabled.")
```
This message doesn't seem particularly useful

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1003,8 +1003,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
-    assert(master.contains("yarn") || dynamicAllocationTesting,
-      "Requesting executors is currently only supported in YARN mode")
+    assert(master.contains("mesos") || master.contains("yarn") || dynamicAllocationTesting,
+      "Requesting executors is currently only supported in YARN or Mesos mode")
```
Mesos coarse-grained mode?

andrewor14 commented:
Hi @tnachen thanks for your work. I will do a closer review within the next few days. First I have a high level question. IIUC, the changes in this PR basically enable the dynamic allocation feature to be used with Mesos coarse-grained mode. However, in Mesos I believe we express the amount of resources an application wants in terms of cores, not the number of executors. Is the concern raised in [SPARK-4751](https://issues.apache.org/jira/browse/SPARK-4751) for standalone mode relevant here?

tnachen committed:
Fix requiring executor resources when executor is still running

SparkQA commented:
[Test build #25206 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25206/consoleFull) for   PR 3861 at commit [`c18760f`](https://github.com/apache/spark/commit/c18760f18af9180a939c6d30a7231c74e17f4db7).
 * This patch merges cleanly.

tnachen commented:
Hi @andrewor14, yes the concerns are identical. Basically in coarse grain mode, I don't over allocate executors even if you request more than the available slaves.
So really dynamic allocation in coarse grain mode simply provides a way to scale down the executors, and when more is needed to scale back maximumly to the available slave nodes and also still within spark.cores.max.

This could change ofcourse as there are discussions to allow coarse grain mode to launch multiple executors, but that's a seperate discussion.

tnachen committed:
Add relaunch task test

SparkQA commented:
[Test build #25209 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25209/consoleFull) for   PR 3861 at commit [`f39cdd2`](https://github.com/apache/spark/commit/f39cdd2fca8f30f57dcbc0c111e6ab147ed7055a).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25206 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25206/consoleFull) for   PR 3861 at commit [`c18760f`](https://github.com/apache/spark/commit/c18760f18af9180a939c6d30a7231c74e17f4db7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SlaveStatus(var executorRunning: Boolean, var taskRunning: Boolean) `
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25206/
Test PASSed.

SparkQA commented:
[Test build #25209 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25209/consoleFull) for   PR 3861 at commit [`f39cdd2`](https://github.com/apache/spark/commit/f39cdd2fca8f30f57dcbc0c111e6ab147ed7055a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SlaveStatus(var executorRunning: Boolean, var taskRunning: Boolean) `
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25209/
Test PASSed.

tnachen committed:
Only deduct shuffle resources when executor is launched.

SparkQA commented:
[Test build #25245 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25245/consoleFull) for   PR 3861 at commit [`99cc898`](https://github.com/apache/spark/commit/99cc898add0e7bcd94f59fb04c26403c557b920e).
 * This patch merges cleanly.

tnachen committed:
Address review feedback

SparkQA commented:
[Test build #25248 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25248/consoleFull) for   PR 3861 at commit [`95ff246`](https://github.com/apache/spark/commit/95ff24636900c105515a3d70cc4424430a7a89e5).
 * This patch merges cleanly.

tnachen commented:
@andrewor14 I updated the PR with your feedback, and also after testing on a 4 node cluster I fixed an issue I found where it the offer has enough resources just for the task but scheduler is looking for shuffle and task resources.

SparkQA commented:
[Test build #25249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25249/consoleFull) for   PR 3861 at commit [`93e5fbc`](https://github.com/apache/spark/commit/93e5fbc7bcfb7a7164b7cfa88ea44289b3ed4710).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25245/consoleFull) for   PR 3861 at commit [`99cc898`](https://github.com/apache/spark/commit/99cc898add0e7bcd94f59fb04c26403c557b920e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SlaveStatus(var executorRunning: Boolean, var taskRunning: Boolean) `
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25245/
Test PASSed.

SparkQA commented:
[Test build #25248 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25248/consoleFull) for   PR 3861 at commit [`95ff246`](https://github.com/apache/spark/commit/95ff24636900c105515a3d70cc4424430a7a89e5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SlaveStatus(var executorRunning: Boolean, var taskRunning: Boolean) `
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25248/
Test FAILed.

SparkQA commented:
[Test build #25249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25249/consoleFull) for   PR 3861 at commit [`93e5fbc`](https://github.com/apache/spark/commit/93e5fbc7bcfb7a7164b7cfa88ea44289b3ed4710).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25249/
Test FAILed.

tnachen commented:
@andrewor14 I wonder if you have time to review this?

SparkQA commented:
[Test build #25581 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25581/consoleFull) for   PR 3861 at commit [`99415c3`](https://github.com/apache/spark/commit/99415c3bc9973f2f80faaf7f5742b3bc860bc900).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25581 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25581/consoleFull) for   PR 3861 at commit [`99415c3`](https://github.com/apache/spark/commit/99415c3bc9973f2f80faaf7f5742b3bc860bc900).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class SlaveStatus(var executorRunning: Boolean, var taskRunning: Boolean) `
  * `      "cd %s*; %s ./bin/spark-class ".format(basename, prefixEnv)`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25581/
Test PASSed.

tnachen commented:
Adding you folks for review: @dragos @deanw @huitseeker @skyluc

andrewor14 commented:
Hey @tnachen sorry again for slipping on this. I think this would be a good feature to add in 1.4. I'll be sure to review this at the beginning of the release cycle. Thanks for pinging me on this.

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+import org.apache.spark.deploy.worker.StandaloneWorkerShuffleService
+import scala.collection.JavaConversions._
+import scala.io.Source
+import java.io.{File, PrintWriter}
```
we should reorder these imports as we do in other files

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+          System.out.println(line)
+        }
+      }
+    }.start()
```
Could you do these through `RedirectThread` in `Utils.scala`?

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+    taskId = taskInfo.getTaskId
+  }
+
+  override def error(d: ExecutorDriver, message: String) {
```
methods without return types should use `: Unit` (see https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide)

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+    driver.sendStatusUpdate(TaskStatus.newBuilder()
+      .setState(TaskState.TASK_RUNNING)
+      .setTaskId(taskInfo.getTaskId)
+      .build)
```
For readability I would prefer that you put the builder in a `val` first, and then calling `driver.sendStatusUpdate` on it. Also, please use parentheses for `build()` here because it's not a getter without side-effects.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -32,6 +32,13 @@ import org.apache.spark.{Logging, SparkContext, SparkEnv, SparkException}
+import org.apache.mesos.protobuf.ByteString
+import java.util.concurrent.locks.ReentrantLock
+
+case class SlaveStatus(var executorRunning: Boolean, var taskRunning: Boolean) {
```
this should not be public. Also, it would be best if you could put this in a different file where all the messages live. Also, why are these vars? Case classes by definition are supposed to have immutable fields.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -193,45 +252,107 @@ private[spark] class CoarseMesosSchedulerBackend(
-            failuresBySlaveId.getOrElse(slaveId, 0) < MAX_SLAVE_FAILURES &&
-            !slaveIdsWithExecutors.contains(slaveId)) {
+        logTrace("Received offer id: " + offer.getId.getValue + ", cpu: " + cpus.toString +
+          ", mem: " + mem.toString)
```
please use string interpolation, here and other places:
```
logTrace(s"Received offer id: ${offer.getId.getValue}, cpu: $cpus, mem: $mem")
```

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -193,45 +252,107 @@ private[spark] class CoarseMesosSchedulerBackend(
+          failuresBySlaveId.getOrElse(slaveId, 0) < MAX_SLAVE_FAILURES &&
+          (!slaveStatuses.contains(slaveId) || !slaveStatuses(slaveId).taskRunning) &&
+          mem >= minMemRequired &&
+          cpus >= minCpuRequired) {
```
this if case is very hard to read. Put a few of these cases in properly named variables and chain them together?

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -193,45 +252,107 @@ private[spark] class CoarseMesosSchedulerBackend(
+    0
+  }
+
+  lazy val shuffleServiceCpu = if (shuffleServiceEnabled) {
```
why are all of these lazy vals? These aren't particularly expensive operations

andrewor14 commented on core/src/test/scala/org/apache/spark/scheduler/mesos/CoarseMesosSchedulerBackendSuite.scala:
```diff
@@ -0,0 +1,210 @@
+import akka.actor.ActorSystem
+import java.util.Collections
+
+class CoarseMesosSchedulerBackendSuite extends FunSuite with LocalSparkContext with EasyMockSugar {
```
Please use Mockito instead of EasyMock. There is actually an outstanding patch to replace all uses of EasyMock with Mockito: https://github.com/apache/spark/pull/4578

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+      "' is registered.")
+
+    if (shuffleService == null) {
+      sparkConf.set("spark.shuffle.service.enabled", "true")
```
We shouldn't set this config ourselves; it is meant to be set by the user. What we should do here is check whether the user set it and then start the shuffle service if so.

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+  private var driver: ExecutorDriver = null
+  private var executorProc: Process = null
+  private var taskId: TaskID = null
+  @volatile var killed = false
```
this should be private, otherwise anyone with a reference to this backend can "kill" it.

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+
+    driver.sendStatusUpdate(TaskStatus.newBuilder()
+      .setState(state)
+      .setMessage(msg)
```
what happens if you set an empty message? Does it still get logged? Should we set the message only if it's not empty?

andrewor14 commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedMesosExecutorBackend.scala:
```diff
@@ -0,0 +1,212 @@
+    // because the backend is designed to run in its own process.
+    // Since it's a shared class we are preserving the existing behavior
+    // and launching it as a subprocess here.
+    val command = "exec " + Utils.deserialize[String](taskInfo.getData().toByteArray)
```
here `command` is a giant string instead of a list of command line arguments. Should we do something like `Utils.splitCommandString` here to preserve the command properly when we pass it to the `ProcessBuilder`?

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/worker/StandaloneWorkerShuffleService.scala:
```diff
@@ -31,7 +31,6 @@ import org.apache.spark.network.shuffle.ExternalShuffleBlockHandler
  * Optionally requires SASL authentication in order to read. See [[SecurityManager]].
  */
-private[worker]
 class StandaloneWorkerShuffleService(sparkConf: SparkConf, securityManager: SecurityManager)
```
The correct thing to do here is not to remove its visibility, but to abstract this class so it is not used exclusively for standalone mode.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -61,22 +68,31 @@ private[spark] class CoarseMesosSchedulerBackend(
-  val slaveIdsWithExecutors = new HashSet[String]
+  val slaveStatuses = new HashMap[String, SlaveStatus]
+
+  val taskIdToSlaveId = HashBiMap.create[Int, String]()
```
why use a special map here instead of just a normal hash map?

andrewor14 commented on docs/running-on-mesos.md:
```diff
@@ -226,6 +226,22 @@ See the [configuration page](configuration.html) for information on Spark config
+  <td>1</td>
+  <td>
+    The amount of CPU that the Mesos coarse grained scheduler will request for launching the shuffle service. The shuffle service is launched on each slave.
+    This will only apply if shuffle service is enabled and is running under coarse grained mode.
```
I'm confused, is this the number of CPUs given to the Mesos slaves or the shuffle service itself? From the description it seems that it's the latter but from the code it seems that it's the former.

tnachen replied:
Mesos slave doesn't need CPU allocated to it, since it was the one that offered the CPU in the first place.
This is the amount of CPU that is required for shuffle service itself. So in the end the total amount of CPU to use is the mesos executor CPU + task CPU + shuffle service CPU.

andrewor14 commented:
@tnachen Thanks for working on this. Before I dive deeper into the implementation there's a main open question that I'd like to address. The external shuffle service is intended to live across the executor lifetime and so launched independently of any Spark application. The service enables dynamic allocation of resources because it can continue to serve an executor's shuffle files after the executor has been killed. However, in this patch the service seems to be started inside the executor backend itself and its fate necessarily tied with the application.

If I understand correctly, the Mesos slave is equivalent to the standalone Worker in that it is long running and lives beyond the lifetime of a particular application. If this is the case, the appropriate place to start the shuffle service would be there instead.

Another issue is that this patch in its current state seems to conflate two issues (1) dynamic allocation and (2) external shuffle service. (1) is what you refer to as auto-scaling on the JIRA, and depends on (2) to work. However, since we already check whether shuffle service is enabled in `ExecutorAllocationManager`, we shouldn't check it again when launching the Mesos executor. More specifically, I don't understand why we launch the executor two different ways depending on whether (2) is enabled. I believe a better solution is to separate these two concerns and launch the executor the same way we already do today.

tnachen commented:
@andrewor14 Mesos slave doesn't have the concept of Auxilary service as YARN, that launches a service as part of the Slave. It requires a task to be running to launch anything.

Therefore to have a long running service that lives longer than the task lifetime, I went the route to create a custom Mesos executor that launches the Coarse grained spark executor, and continue to run when the spark executor finishes. The custom executor also launches the external shuffle service.

I agree I didn't really like the diverging path, but I felt like the original method is much simpler as it directly launches the coarse grain executor, and don't want to increase the complexity by making it also have to launch this custom executor.

pwendell commented:
We spoke a bit offline about this, but my feeling was that the best thing here might be to add a way to launch the shuffle service as a standalone application (initially, not one managed by Mesos) so that it can be shared across Spark applications. That would involve writing some simple launching scripts for it in a similar way to existing daemons we launch, and you'd ask users to launch the shuffle service similar to other storage systems like HDFS. That's very simple and would avoid diverging a lot between Mesos and the other modes. And longer term we could actually have a single shared shuffle service that is scheduled by mesos.

tnachen commented:
Agree and it's currently being worked on. We can close this PR too.

dragos commented:
I'd like to take a shot at making the external shuffle service a "daemon". Just to make sure I understand the requirements:

- the shuffle service is launched manually on cluster machines
- it does not care about `spark.shuffle.service.enabled`, if it's started -- it just starts
- it should not depend on Spark or Scala -- this is the part I'd like to get confirmation. If that's the case, it means we can't use helper scripts like `spark-submit` or `spark-class` that would simplify loading configuration files

/cc @tnachen
